dhall/src/Dhall/Lint.hs

```diff
@@ -26,11 +27,17 @@ lint = Dhall.Optics.rewriteOf subExpressions removeUnusedBindings . removeLetInL
 removeUnusedBindings :: Eq a => Expr s a -> Maybe (Expr s a)
 -- Don't remove assertions!
 removeUnusedBindings (Let _ _ (Assert _) _) = Nothing
+removeUnusedBindings (Let _ _ e _) | isAssertInLam e = Nothing
 removeUnusedBindings (Let a _ _ d)
     | not (V a 0 `Dhall.Core.freeIn` d) =
         Just (Dhall.Core.shift (-1) (V a 0) d)
 removeUnusedBindings _ = Nothing
 
+isAssertInLam :: Expr s a -> Bool
+isAssertInLam (Lam _ _ (Assert _)) = True
+isAssertInLam (Lam _ _ e) = isAssertInLam e
```

Comment:
I think this check isn't sufficiently reliable because you might have an expression like:

```haskell
λ(m : Natural) → let n = m + 0 in assert : m ≡ n
```

What you probably want to do is to use `subExpressions` like this to catch all possible occurrences of `Assert` within the sub-expression, like this:

```haskell
isAssertInLam e = not (null [ Assert _ <- toListOf subExpressions e ])
```

Note that I haven't tested if that actually works, but you'll see several examples of that idiom in `dhall-lsp-server`

Side note: this is out of scope for this pull request, but if we had `Prism`s for each constructor ,then we could simplify `isAssertInLam` even more to:

```
isAssertInLam = Lens.has (subExpressions . _Assert)
```